### ISSUE TITLE
Fix pallet error spam when there is really just one simple error

### ISSUE THREAD

**Issue by sam0x17** [paritytech/substrate#13432]
Created from this comment https://github.com/paritytech/substrate/issues/12011#issuecomment-1438845701

> Currently, if your pallet has a syntax error, you get an error for it, and then a gzillion other errors because the corrupt/incomplete pallet code is still being treated as Rust code. Ideally, we'd tell the compiler to stop immediately.

This is largely due to missing trait impls or structs during pallet creation resulting from some error that prevented some part of the expansion. There are a lot of situations where this can happen, so the surface area for fixing this issue could be rather large, however the approach should be to proactively detect that something is missing and issue a compile error there, _before_ pallet expansion begins.

Generally speaking, we want to make sure that the "validation" phase of all of our pallet macros happens during parsing, and that avoid issuing errors during expansion

**Comment by bkchr** [comment]
I also thought about this already multiple times. If we introduce some validation phase, which we generally already have as the parsing phase or whatever we do, we should always expand. Maybe if we detect issues we expand with some "best effort". Otherwise we would still see a lot of unrelated errors.

But yeah, in general that is probably something that requires some tinkering to find the best way to handle certain errors.

**Comment by sam0x17** [comment]
Yeah, I think the key is to catch these types of errors and issue a compiler error _before_ any expansion has started, but will have to try it to know for sure

**Comment by bkchr** [comment]
I don't think that we currently generate types on best effort. Afaik we bubble up the error to the top and then don't generate anything.
